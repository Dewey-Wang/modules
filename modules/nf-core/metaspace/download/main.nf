process METASPACE_DOWNLOAD {
    label 'process_low'

    publishDir 'results', mode: 'copy'
    container 'bwadie/metaspace_converter'

    input:
    tuple val(dataset_id), val(database), val(version)
    path script_file

    output:
    path "${dataset_id}_*.csv", optional: true, emit: results  // 設為可選
    stdout emit: log  // 將 dataset_id 與 stdout 一起輸出

    script:
    """
    python3 $script_file \\
        --dataset_id "$dataset_id" \\
        --database "${database ?: 'None'}" \\
        --version "${version ?: 'None'}"
    """
}

workflow {
    script_file = file("${projectDir}/metaspace_download.py")

    if (params.datasets_file) {
        datasets = Channel.fromPath(params.datasets_file)
            .splitCsv(header: true, strip: true)
            .map { row ->
                // 將空值轉換為 null
                def database = row.database?.trim() ?: null
                def version = row.version?.trim() ?: null
                [row.dataset_id, database, version]
            }
    } else {
        datasets = Channel.fromList(params.datasets)
    }

    METASPACE_DOWNLOAD(datasets, script_file)
    METASPACE_DOWNLOAD.out.log.view { "${it.split('\n').last().trim()}" }
}
