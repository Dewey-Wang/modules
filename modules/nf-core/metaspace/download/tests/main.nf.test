// nf-core modules test metaspace/download
nextflow_process {

    name "Test Process METASPACE_DOWNLOAD"
    script "../main.nf"
    process "METASPACE_DOWNLOAD"

    tag "modules"
    tag "modules_nfcore"
    tag "metaspace"
    tag "metaspace/download"

    test("METASPACE_DOWNLOAD - datasets") {
        when {
            process {
                """
                input[0] = Channel.from([
                    ['2022-08-05_17h28m56s', 'KEGG', null],
                    ['2022-08-05_17h28m56s', null, null]
                ])
                params.output = 'test_results' // Override default output directory for testing
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert file("${process.out.results[0]}").exists() } // Check if a CSV file was generated
            )
        }
    }

    test("METASPACE_DOWNLOAD - wrong_format") {
        when {
            process {
                """
                input[0] = Channel.from([
                    ['2022-08-05_17h28m56s', 'KEGG', null],
                    ['2022-08-05_17h28m56s', null, null],
                    [null, 'KEGG', null] // Missing dataset_id
                ])
                params.output = 'test_results' // Override default output directory for testing
                """
            }
        }

        then {
            assertAll(
                { assert process.failed } // Process may succeed unless metaspace_download.py fails on null
            )
        }
    }
}