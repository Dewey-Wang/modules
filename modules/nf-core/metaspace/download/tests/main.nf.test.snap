{
    "sarscov2 - bam - stub": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "bam": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-25T11:48:49.633583"
    },
    "sarscov2 - bam": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "bam": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-25T11:48:26.180059"
    }
}